spring-rabbit/src/test/java/org/springframework/amqp/rabbit/listener/DirectMessageListenerContainerMockTests.java

```diff
@@ -96,4 +102,104 @@ public void testAlwaysCancelAutoRecoverConsumer() throws Exception {
+		willAnswer(i -> {
+			consumer.set(i.getArgument(6));
+			consumer.get().handleConsumeOk("consumerTag");
+			latch1.countDown();
+			return "consumerTag";
+		})
+		.given(channel).basicConsume(anyString(), anyBoolean(), anyString(), anyBoolean(), anyBoolean(),
+						anyMap(), any(Consumer.class));
+
+		final AtomicInteger qos = new AtomicInteger();
+		willAnswer(i -> {
+			qos.set(i.getArgument(0));
+			return null;
+		}).given(channel).basicQos(anyInt());
+		final CountDownLatch latch2 = new CountDownLatch(2);
+		final CountDownLatch latch3 = new CountDownLatch(1);
+		willAnswer(i -> {
+			if (i.getArgument(0).equals(10L) || i.getArgument(0).equals(16L)) {
+				latch2.countDown();
+			}
+			else if (i.getArgument(0).equals(17L)) {
+				latch3.countDown();
+			}
+			return null;
+		}).given(channel).basicAck(anyLong(), anyBoolean());
+		final CountDownLatch latch4 = new CountDownLatch(1);
+		willAnswer(i -> {
+			latch4.countDown();
+			return null;
+		}).given(channel).basicNack(19L, true, true);
+
+		DirectMessageListenerContainer container = new DirectMessageListenerContainer(connectionFactory);
+		container.setQueueNames("test");
+		container.setPrefetchCount(2);
+		container.setMonitorInterval(100);
+		container.setMessagesPerAck(10);
+		container.setAckTimeout(100);
+		container.setMessageListener(m -> {
+			if (m.getMessageProperties().getDeliveryTag() == 19L) {
+				throw new RuntimeException("TestNackAndPendingAcks");
+			}
+		});
+		container.afterPropertiesSet();
+		container.start();
+
+		assertTrue(latch1.await(10, TimeUnit.SECONDS));
+		assertThat(qos.get(), equalTo(2));
+		BasicProperties props = new BasicProperties();
+		byte[] body = new byte[1];
+		for (long i = 1; i < 16; i++) {
+			consumer.get().handleDelivery("consumerTag", envelope(i), props, body);
+		}
+		Thread.sleep(200);
+		consumer.get().handleDelivery("consumerTag", envelope(16), props, body);
+		// should get 2 acks #10 and #6 (timeout)
+		assertTrue(latch2.await(10, TimeUnit.SECONDS));
+		consumer.get().handleDelivery("consumerTag", envelope(17), props, body);
+		verify(channel).basicAck(10L, true);
+		verify(channel).basicAck(16L, true);
+		assertTrue(latch3.await(10, TimeUnit.SECONDS));
+		// monitor task timeout
+		verify(channel).basicAck(17L, true);
+		consumer.get().handleDelivery("consumerTag", envelope(18), props, body);
+		consumer.get().handleDelivery("consumerTag", envelope(19), props, body);
+		assertTrue(latch4.await(10, TimeUnit.SECONDS));
+		// pending acks before nack
+		verify(channel).basicAck(18L, true);
+		verify(channel).basicNack(19L, true, true);
+		consumer.get().handleDelivery("consumerTag", envelope(20), props, body);
+		final CountDownLatch latch5 = new CountDownLatch(1);
+		willAnswer(i -> {
+			consumer.get().handleCancelOk("consumerTag");
+			latch5.countDown();
+			return null;
+		}).given(channel).basicCancel("consumerTag");
+		Executors.newSingleThreadExecutor().execute(() -> container.stop());
```

Comment:
My IDEA suggest me to do this: `container::stop`.
Not a big deal. Just in case we still pursue the holy war for the best IDE 😉